File: java-to-go-projects/17-kv-store/Store.java
package kvstore;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

/**
 * In-memory key-value store with thread-safe access.
 * Supports basic CRUD operations and key listing.
 */
public class Store {

    private final ConcurrentHashMap<String, String> data = new ConcurrentHashMap<>();

    public String get(String key) {
        return data.get(key);
    }

    public void set(String key, String value) {
        data.put(key, value);
    }

    public boolean delete(String key) {
        return data.remove(key) != null;
    }

    public Set<String> keys() {
        return data.keySet();
    }

    public int size() {
        return data.size();
    }

    /**
     * Replace all data with the given map (used by Persistence.load).
     */
    public void replaceAll(Map<String, String> newData) {
        data.clear();
        data.putAll(newData);
    }

    /**
     * Get a snapshot of all data (used by Persistence.save).
     */
    public Map<String, String> snapshot() {
        return new HashMap<>(data);
    }
}

File: java-to-go-projects/17-kv-store/rust/src/store.rs
use std::collections::HashMap;

pub struct Store {
    data: HashMap<String, String>,
}

impl Store {
    pub fn new() -> Self{
        Store {
            data: HashMap::new(),
        }
    }
    pub fn get(&self, key: &str) -> Option<&String>{
        self.data.get(key)
    }
    pub fn set(&mut self, key: String, value: String){
        self.data.insert(key, value);
    }
    pub fn delete(&mut self, key: &str) -> bool{
        self.data.remove(key).is_some()
    }
    pub fn keys(&self) -> Vec<&String>{
        self.data.keys().collect()
    }
    pub fn size(&self) -> usize{
        self.data.len()
    }
    pub fn snapshot(&self) -> HashMap<String, String>{
        self.data.clone()
    }
    pub fn replace_all(&mut self, data: HashMap<String, String>){
        self.data = data;
    }
}